File: MainProject/Code.Infras.Writers.Abstractions/PlanetGenerates/IPointRepo.cs
using Commons.Utils.HexSphereGrid;
using Domains.Models.Entities.PlanetGenerates;
using Godot;
using Infras.Writers.Abstractions.Bases;

namespace Infras.Writers.Abstractions.PlanetGenerates;

public interface IPointRepo : IRepository<Point>
{
    Point Add(bool chunky, Vector3 position, SphereAxial coords);
    Point? GetByCoords(bool chunky, SphereAxial coords);
    int? GetIdByCoords(bool chunky, SphereAxial coords);
    Point? GetByPosition(bool chunky, Vector3 position);
    int? GetIdByPosition(bool chunky, Vector3 position);
    IEnumerable<Point> GetAllByChunky(bool chunky);
    List<Point> GetNeighborCenterIds(List<Face> hexFaces, Point center);
    SphereAxial GetSphereAxial(Tile tile);
}

File: TerraObserver/TO.Infras.Abstractions/Planets/Repos/IPointRepo.fs
namespace TO.Infras.Abstractions.Planets.Repos

open Friflo.Engine.ECS
open Godot
open TO.Commons.Structs.HexSphereGrid
open TO.Infras.Abstractions.Planets.Models.Faces
open TO.Infras.Abstractions.Planets.Models.Points

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-05-28 09:21:28
[<Interface>]
type IPointRepo =
    abstract ForEachByChunky: chunky: bool -> forEachPoint: (PointComponent -> Entity -> unit) -> unit
    abstract TryHeadByPosition: chunky: bool -> pos: Vector3 -> Entity option
    abstract Add: chunky: bool -> position: Vector3 -> coords: SphereAxial -> int

    abstract GetNeighborCenterPointIds:
        chunky: bool * hexFaces: FaceComponent list * center: PointComponent inref -> int ResizeArray

    abstract CreateVpTree: chunky: bool -> unit
    abstract SearchNearestCenterPos: pos: Vector3 * chunky: bool -> Vector3
    abstract Truncate: unit -> unit
